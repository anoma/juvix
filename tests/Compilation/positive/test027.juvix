--- Church numerals
--- This test is disabled until https://github.com/anoma/juvix/issues/1706 is fixed.
--- It is convenient to comment it out so we can format all test files without a crash.
module test027;

-- import Stdlib.Prelude open hiding {toNat};

-- Num : Type := {A : Type} → (A → A) → A → A;

-- czero : Num
--   | {_} f x := x;

-- csuc : Num → Num
--   | n {_} f := f << n {_} f;

-- num : Nat → Num
--   | zero := czero
--   | (suc n) := csuc (num n);

-- add : Num → Num → Num
--   | n m {_} f := n {_} f << m {_} f;

-- mul : Num → Num → Num
--   | n m {_} := n {_} << m {_};

-- isZero : Num → Bool
--   | n := n {_} (const false) true;

-- toNat : Num → Nat
--   | n := n {_} ((+) 1) 0;

-- main : IO :=
--   printNatLn (toNat (num 7))
--     >>> printNatLn (toNat (add (num 7) (num 3)))
--     >>> printNatLn (toNat (mul (num 7) (num 3)));
