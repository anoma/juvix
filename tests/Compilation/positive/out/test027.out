7
10
21
